### ISSUE TITLE
CA2000 false positve when combining ValueTask and local function

### ISSUE THREAD

**Issue by Tan90909090** [dotnet/roslyn-analyzers#3516]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v3.0.0-beta3.final (latest pre-release)

#### Diagnostic ID

[CA2000](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000)

#### Repro steps
```cs
using System.IO;
using System.Threading.Tasks;

public sealed class C
{
    public static ValueTask<Stream> SomeAsync()
    {
        return CoreAsync();

        async ValueTask<Stream> CoreAsync()
        {
            await Task.CompletedTask.ConfigureAwait(false);
            return new MemoryStream(); // CA2000 occurs here
        }
    }

    private static void Main() { }
}
```

#### Expected behavior
CA2000 does not occur because `SomeAsync` and `CoreAsync` are intended for returning an instance.

#### Actual behavior
CA2000 occurs: `Call System.IDisposable.Dispose on object created by 'new MemoryStream()' before all references to it are out of scope.`

#### Note
If I don't use a local function then CA2000 does NOT occur:
```cs
public async static ValueTask<Stream> SomeAsync()
{
    await Task.CompletedTask.ConfigureAwait(false);
    return new MemoryStream(); // CA2000 does NOT occur
}
```
In addition, if i use `Task` insted of `ValueTask` then CA2000 does NOT occur:
```cs
public static Task<Stream> SomeAsync()
{
    return CoreAsync();

    async Task<Stream> CoreAsync()
    {
        await Task.CompletedTask.ConfigureAwait(false);
        return new MemoryStream(); // CA2000 does NOT occur
    }
}
```

**Comment by Evangelink** [comment]
I confirm the various behavior described by @Tan90909090.

@mavasani I have tried to debug the 3 cases explained and there are 2 things I can't understand:

1. I can't find the piece of code that actually exclude the disposable value returned by a function. From the debugging, there is nothing that seems to be saying even if this isn't disposed don't report.

2. Maybe this question will be answered when I see how we do the exclusion but I don't understand why we don't exclude all disposable items being returned from a method.

**Comment by mavasani** [comment]
> I don't understand why we don't exclude all disposable items being returned from a method.

That is a primary case of dispose leaks from factory methods. We special case certain methods as known factory methods and track dispose state for them.

I will try to look at this issue this week.
